swift/extractor/remapping/SwiftFileInterception.cpp

```diff
@@ -0,0 +1,169 @@
+#include "swift/extractor/remapping/SwiftFileInterception.h"
+
+#include <fcntl.h>
+#include <filesystem>
+
+#include <dlfcn.h>
+#include <mutex>
+#include <optional>
+#include <cassert>
+
+#include "swift/extractor/infra/file/FileHash.h"
+#include "swift/extractor/infra/file/FileHash.h"
+
+#ifdef __APPLE__
+// path is hardcoded as otherwise redirection could break when setting DYLD_FALLBACK_LIBRARY_PATH
+#define SHARED_LIBC "/usr/lib/libc.dylib"
+#define FILE_CREATION_MODE O_CREAT
+#else
+#define SHARED_LIBC "libc.so.6"
+#define FILE_CREATION_MODE (O_CREAT | O_TMPFILE)
+#endif
+
+namespace fs = std::filesystem;
+
+namespace {
+namespace original {
+
+void* openLibC() {
+  if (auto ret = dlopen(SHARED_LIBC, RTLD_LAZY)) {
+    return ret;
+  }
+  std::cerr << "Unable to dlopen " SHARED_LIBC "!\n";
+  std::abort();
+}
+
+void* libc() {
+  static auto ret = openLibC();
+  return ret;
+}
+
+template <typename Signature>
+Signature get(const char* name) {
+  return reinterpret_cast<Signature>(dlsym(libc(), name));
+}
+
+int open(const char* path, int flags, mode_t mode = 0) {
+  static auto original = get<int (*)(const char*, int, ...)>("open");
+  return original(path, flags, mode);
+}
+
+}  // namespace original
+
+bool endsWith(const std::string_view& s, const std::string_view& suffix) {
+  return s.size() >= suffix.size() && s.substr(s.size() - suffix.size()) == suffix;
+}
+
+auto& fileInterceptorInstance() {
+  static std::weak_ptr<codeql::FileInterceptor> ret{};
+  return ret;
+}
+
+bool mayBeRedirected(const char* path, int flags = O_RDONLY) {
+  return (!fileInterceptorInstance().expired() && (flags & O_ACCMODE) == O_RDONLY &&
+          endsWith(path, ".swiftmodule"));
+}
+
+}  // namespace
+
+namespace codeql {
+
+class FileInterceptor {
+ public:
+  FileInterceptor(fs::path&& workingDir) : workingDir{std::move(workingDir)} {
+    fs::create_directories(hashesPath());
+    fs::create_directories(storePath());
+  }
+
+  int open(const char* path, int flags, mode_t mode = 0) const {
+    fs::path fsPath{path};
+    assert((flags & O_ACCMODE) == O_RDONLY);
+    errno = 0;
+    // first, try the same path underneath the artifact store
```

Comment:
Using the path underneath the artifact store is in essence a persistent cache of path -> hash mappings. 
How robust is this to file moves (and overrides) made by the build system (i.e. invisible to the extractor). 

I understand that this most likely does not happen, but if it does I gather it will be quite hard to debug.

Reply:
Btw. What is the performance penalty if you disable the first step here, and always recompute the file hash? Was that the cause of the x6 slowdown observed on MacOS?

Reply:
If a file content changes retaining the same path, this will in effect not work. While this kind of optimization here might be avoidable, I found it to be unavoidable in my linkage awareness work, or otherwise extraction on macOS becomes much, too much slower.

One thing I can try, is whether it would suffice to limit the persistence of this path -> hash mapping to just the single extractor run (by keeping it in memory). That would be safe, and maybe performance-saving enough

Reply:
to be precise: this is robust to file moves to new locations: when trying to read a moved file, the path -> hash mapping will not give anything, then we will fall back on computing the hash, and we will find the artifact in the hash -> file mapping.

Overrides on the other hand is what this would break on, that is when the same path is used during the build with different contents. Sane build systems (including SPM and I think xcode) don't exhibit this, but it remains a possibility in a less sane project (which however is much less probable in Swift than in C/C++).

Reply:
> I found it to be unavoidable in my linkage awareness work, or otherwise extraction on macOS becomes much, too much slower.

Do you mean unavoidable due to performance or correctness?

Reply:
perfomance. But I need some more testing.

Reply:
It would be nice to understand what the performance hit introduced by doing open redirection, and also understand if the open redirection is CPU or IO bound to allow us to understand potential speedup if we want to speed up extraction later.